agora/Agora/Proposal.hs

```diff
@@ -181,24 +254,45 @@ instance PUnsafeLiftDecl PResultTag where type PLifted PResultTag = ResultTag
 deriving via
   (DerivePConstantViaNewtype ResultTag PResultTag PInteger)
   instance
-    (PConstant ResultTag)
+    (PConstantDecl ResultTag)
+
+-- FIXME: This instance and the one below, for 'PProposalId', should be derived.
+-- Soon this will be possible through 'DerivePNewtype'.
+instance PTryFrom PData (PAsData PResultTag) where
+  type PTryFromExcess PData (PAsData PResultTag) = PTryFromExcess PData (PAsData PInteger)
```

Comment:
I cannot recall seeing `PTryFromExcess` before. Could you fill me in?

Reply:
Okay, so. `PTryFrom` the class exists to "safely" cast from an unknown type to a known type. In order to do this, it needs to verify that the shape is valid. It traverses the data (e.g. using `pfstBuiltin` and `pasList`, etc). `PTryFromExcess` is then the witness of the proof that the input is actually of that shape, and the "result" of traversing the data. If the data Sometimes, there is no proof retrieved, for example newtypes, (Which cannot be checked on-chain!) in these situations we use `Const ()`.